src/cmd/partition.rs

```diff
@@ -0,0 +1,223 @@
+
+impl FilenameTemplate {
+    /// Generate a new filename using `unique_value` to replace the `"{}"`
+    /// in the template.
+    fn filename(&self, unique_value: &str) -> String {
+        format!("{}{}{}", &self.prefix, unique_value, &self.suffix)
+    }
+
+    /// Create a new, writable file in directory `path` with a filename
+    /// using `unique_value` to replace the `"{}"` in the template.  Note
+    /// that we do not output headers; the caller must do that if
+    /// desired.
+    fn writer<P>(&self, path: P, unique_value: &str) -> io::Result<BoxedWriter>
+        where P: AsRef<Path>
+    {
+        let filename = self.filename(unique_value);
+        let full_path = path.as_ref().join(filename);
+        let spath = Some(full_path.display().to_string());
+        Config::new(&spath).writer()
+    }
+}
+
+impl Decodable for FilenameTemplate {
+    fn decode<D: Decoder>(d: &mut D) -> Result<FilenameTemplate, D::Error> {
+        let raw = try!(d.read_str());
+        let chunks = raw.split("{}").collect::<Vec<_>>();
+        if chunks.len() == 2 {
+            Ok(FilenameTemplate {
+                prefix: chunks[0].to_owned(),
+                suffix: chunks[1].to_owned(),
+            })
+        } else {
+            Err(d.error("The --filename argument must contain one '{}'."))
+        }
+    }
+}
+
+/// Generates unique filenames based on CSV values.
+struct WriterGenerator {
+    template: FilenameTemplate,
+    counter: usize,
+    used: HashSet<String>,
+    non_word_char: Regex,
+}
+
+impl WriterGenerator {
+    fn new(template: FilenameTemplate) -> WriterGenerator {
+        WriterGenerator {
+            template: template,
+            counter: 1,
+            used: HashSet::new(),
+            non_word_char: Regex::new(r#"\W"#).unwrap(),
+        }
+    }
+
+    /// Create a CSV writer for `key`.  Does not add headers.
+    fn writer<P>(&mut self, path: P, key: &csv::ByteString)
+                 -> io::Result<BoxedWriter>
+        where P: AsRef<Path>
+    {
+        let unique_value = self.unique_value(key);
+        self.template.writer(path.as_ref(), &unique_value)
+    }
+
+    /// Generate a unique value for `key`, suitable for use in a
+    /// "shell-safe" filename.  If you pass `key` twice, you'll get two
+    /// different values.
+    fn unique_value(&mut self, key: &csv::ByteString) -> String {
+        // Sanitize our key.
+        let utf8 = String::from_utf8_lossy(key);
+        let safe = self.non_word_char.replace_all(&*utf8, "");
+        let base = if safe.is_empty() {
+            "empty".to_owned()
+        } else {
+            safe
+        };
+
+        // Now check for collisions.
+        if !self.used.contains(&base) {
+            self.used.insert(base.clone());
+            base
+        } else {
+            // This will hang indefinitely if we somehow manage to use all
+            // possible candidates, but to do this we would need to create
+            // more than 2^64 open files.
```

Comment:
Maybe set a reasonable upper bound and fail with an error message? (If there is something the user can do to increase their chances of success, the error message should mention what that is.)